scripts/untar_gz_files.py

```diff
@@ -62,25 +61,66 @@ def untar_gz_files(
+    Uploads and then deletes each extracted file, one by one.
+    Deletes the original downloaded tarball after all its files are uploaded.
+
+    The uploaded files end up in a directory denoted by the destination
+    appended to the original gs:// search path.
     """
     input_bucket = client.get_bucket(bucket_name)
 
     for blob_name in blob_names:
-        input_blob = input_bucket.get_blob(blob_name).download_as_string()
-        with tarfile.open(fileobj=io.BytesIO(input_blob)) as tar:
-            logging.info(f'Untaring {blob_name}')
-            for member in tar.getnames():
-                tar.extract(member, path=f'./{blob_name}')
-                output_blob = input_bucket.blob(
-                    os.path.join(subdir, destination, member)
-                )
-                output_blob.upload_from_filename(f'./{blob_name}/{member}')
-                logging.info(
-                    f'{member} extracted to gs://{bucket_name}/{subdir}/{destination}/{member}'
-                )
+        # Make the extraction directory on the disk
+        if not os.path.exists(f'./{subdir}'):
+            os.makedirs(f'./{subdir}')
+            os.makedirs(f'./{subdir}/extracted')
+
+        # Download and extract the tarball
+        input_bucket.get_blob(blob_name).download_to_filename(blob_name)
+        logging.info(f'Untaring {blob_name}')
+        subprocess.run(
+            ['tar', '-xzf', f'{blob_name}', '-C', f'./{subdir}/extracted'],
+            check=True,
+        )
+        logging.info(f'Untared {blob_name}')
+
+        extracted_from_tarball = os.listdir(f'./{subdir}/extracted')
+
+        # Check if the tarball compressed a single directory, if yes then get files inside
+        if os.path.isdir(f'./{subdir}/extracted/{extracted_from_tarball[0]}'):
+            is_directory = True
+            folder = extracted_from_tarball[0]
+            extracted_files = os.listdir(f'./{subdir}/extracted/{folder}')
+        else:
+            is_directory = False
+            extracted_files = os.listdir(f'./{subdir}/extracted')
+        logging.info(f'Extracted {extracted_files}')
+
+        # Iterate through extracted files, upload them to bucket, then delete them
+        for file in extracted_files:
+            output_blob = input_bucket.blob(os.path.join(subdir, destination, file))
+
+            if is_directory:
+                filepath = f'./{subdir}/extracted/{folder}/{file}'
+            else:
+                filepath = f'./{subdir}/extracted/{file}'
+
+            output_blob.upload_from_filename(filepath)
+            logging.info(
+                f'Uploaded {file} to gs://{bucket_name}/{subdir}/{destination}/'
+            )
+
+            # Delete file after upload
+            subprocess.run(['rm', f'{filepath}'], check=True)
+            logging.info(f'Deleted {file} from disk')
+
+        # Delete tarball after all extracted files uploaded & deleted
+        subprocess.run(['rm', f'{blob_name}'], check=True)
```

Comment:
You might want to make this optional. I often would not want to rm the originals until I am 100% confident everything down stream has gone well.

Reply:
That's fair, I added this part because I'm concerned about disk space for the hail batch job. 

Since we download the tarball and extract it into disk, we might be holding up to ~180GB on disk for the biggest tarballs (90GB tarball + 90GB extracted files). I'm not sure if the hail batch job will even have enough storage capacity to complete this, even with deletions each iteration.

Also noting this wouldn't delete the tarball from the bucket, just from the hail batch disk.